### PR TITLE
Fail sni.yaml loading if related resources fail to load

### LINKED ISSUE #9093: 9.2: mTLS client key load failure should revert to last known good config
These errors were logged after failing to load an mTLS key in ATS 9.2: 

```
ERROR: SSL::139826296874752:error:0B080074:x509 certificate routines:X509_check_private_key:key values mismatch:crypto/x509/x509_cmp.c:303
ERROR: failed to use client private key file from <path>
ERROR: failed to create SSL client session
```

Then later: 

```
CONNECT: attempt fail [CONNECTION_ERROR] to <ip>:<port> for host=<Host> connection_result=Unknown error 524289 [524289] error=Unknown error 524289 [524289] attempts=<connect_attemps_max_retries> url=<url>
```

Resulting in 502s sent downstream. Expected behavior is to revert to last known good config, I would think.

### PR DESCRIPTION
**Pull request by randall**
Fixes #9093
    
Co-authored-by: Chris McFarlen <cmcfarlen@apple.com>

### CODE REVIEW

**Comment by randall** [comment]
This should go into 9.2 and probably others, but they'll require version specific PRs due to slight source reorg/rename.

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
Looks like there is a hard fail for Au test tls_check_cert_selection_reload.

**Comment by randall** [comment]
> Looks like there is a hard fail for Au test tls_check_cert_selection_reload.

Thanks, I think I got it happy now.

**Review by bneradt — CHANGES_REQUESTED** [review]
Looks good to me.

Do you think we can add an autest to verify this? Alan wrote some tests for remap reload failure with the following PR, in case that's helpful:

https://github.com/apache/trafficserver/pull/8802/files

**Comment by randall** [comment]
> Looks good to me.
> 
> Do you think we can add an autest to verify this? Alan wrote some tests for remap reload failure with the following PR, in case that's helpful:
> 
> https://github.com/apache/trafficserver/pull/8802/files

Added (thanks to @cmcfarlen !)

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+)
+
+ts.Disk.sni_yaml.AddLines([
+    "sni:",
+    f"- fqdn: {sni_domain}",
+    "  http2: off",
+    f"  client_cert: {ts.Variables.SSLDir}/signed-foo.pem",
+    f"  client_key: {ts.Variables.SSLDir}/signed-foo.key",
+    "  verify_client: STRICT",
+])
```
Comment (bneradt) [inline tests/gold_tests/tls/tls_sni_yaml_reload.test.py:69]: `f"""` with `.split('\n')`:

```
ts.Disk.sni_yaml.AddLines(
      f"""                                    
      sni:                                    
      - fqdn: {sni_domain}                                    
        http2: off                                    
        client_cert: {ts.Variables.SSLDir}/signed-foo.pem                                    
        client_key: {ts.Variables.SSLDir}/signed-foo.key                                    
        verify_client: STRICT                                    
      """).split('\n')                                    
 )

**Review by bneradt — CHANGES_REQUESTED** [review]
Looks great. Thank you for working on the test. I just have a few minor updates for it.

**Comment by randall** [comment]
oops, I didn't release you added more comments @bneradt. Sorry for the re-review request.

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import os
+
+
+Test.Summary = '''
+Test reloading sni.yaml behaves as expected
+'''
+
+sni_domain = 'example.com'
+
+ts = Test.MakeATSProcess("ts", command="traffic_manager", select_ports=True, enable_tls=True)
```
[the remark this answers is not in the document]
Comment (randall) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py:26]: We should go through and cleanup the 119 other `select_ports=True` in a separate PR too (I (and probably others) just copy/paste from existing tests to make the system happy)

[the remark this answers is not in the document]
Comment (randall) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py]: Created #9142 for select_ports cleanup

**Comment by randall** [comment]
[approve ci autest]

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bneradt) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py]: Thanks Randall. Approved.

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+trupd.Processes.Default.ReturnCode = 0
+
+
+tr2reload = Test.AddTestRun("Reload config")
+tr2reload.StillRunningAfter = ts
+tr2reload.StillRunningAfter = server
+tr2reload.Processes.Default.Command = 'traffic_ctl config reload'
+tr2reload.Processes.Default.Env = ts.Env
+tr2reload.Processes.Default.ReturnCode = 0
+#ts.Disk.diags_log.Content = Testers.ContainsExpression('ERROR: ', 'ERROR')
```
[the remark this answers is not in the document]
Comment (randall) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py]: restored.

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+tr3 = Test.AddTestRun(f"Make request again for {sni_domain} that should still work")
+# Wait for the reload to complete
+#tr3.Processes.Default.StartBefore(server2, ready=When.FileContains(ts.Disk.diags_log.Name, 'unknown value "BROKEN"', 1))
```
[the remark this answers is not in the document]
Comment (randall) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py]: added

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+tr.Processes.Default.StartBefore(Test.Processes.ts)
+tr.Processes.Default.StartBefore(server)
+tr.StillRunningAfter = ts
+tr.StillRunningAfter = server
+tr.Processes.Default.Command = f"curl -q --tls-max 1.2 -s -v -k  --cert ./signed-foo.pem --key ./signed-foo.key --resolve '{sni_domain}:{ts.Variables.ssl_port}:127.0.0.1' https://{sni_domain}:{ts.Variables.ssl_port}"
+
+tr.Processes.Default.ReturnCode = 0
+tr.Processes.Default.Streams.stdout = Testers.ExcludesExpression("Could Not Connect", "Check response")
+tr.Processes.Default.Streams.stderr = Testers.IncludesExpression(f"CN={sni_domain}", "Check response")
+ts.Disk.diags_log.Content = Testers.IncludesExpression("SSL negotiation finished successfully", "negotiation failed")
```
[the remark this answers is not in the document]
Comment (randall) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py:83]: added

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py (hunk 1 above)
Comment (randall) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py]: done in both places.

File: tests/gold_tests/tls/tls_sni_yaml_reload.test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bneradt) [reply tests/gold_tests/tls/tls_sni_yaml_reload.test.py:83]: Thanks, the first two looks good. But can you change the `"negotiation failed"` to something like `"Verify that the TLS handshake was successful."`.

As it is, if the match failed, the failure message will look something like:

```
       file <some_path>/diags_log : negotiation failed - Failed
```

That would look like it's saying we wanted the negotiation to fail, but instead it failed to fail.

Anyway, all the other Expression strings in the test look fine.

**Review by bneradt — APPROVED** [review]
Looks good to me.

**Comment by zwoop** [comment]
Cherry-picked to v9.2.x

**Pull request merged**